// nodejs-randomnames/NameGenerator.js
var NameGenerator = function(){
	this.maleNames = require('./maleNames.js');
	this.femaleName = require('./femaleNames.js');
	this.names = require('./maleNames.js').concat(require('./femaleNames.js'));
};

NameGenerator.prototype.getRandomName = function(gender){
	return this.names[Math.floor(Math.random()*this.names.length)];
};

NameGenerator.prototype.getRandomNames = function(number){
	var randomNames = [];
	for (var i = 0; i < number; i++) {
		var randomName = this.names[Math.floor(Math.random()*this.names.length)];
		randomNames.push(randomName);
	}

	return randomNames;
};

NameGenerator.prototype.getFullName = function(){
	var randomNames = [];

	for (var i = 0; i < 3; i++) {
		var randomName = this.names[Math.floor(Math.random()*this.names.length)];
		randomNames.push(randomName);
	}

	return randomNames.join(" ");
};

NameGenerator.prototype.getFullNameObject = function(){
	var fullName = {
		name: this.names[Math.floor(Math.random()*this.names.length)],
		firstName: this.names[Math.floor(Math.random()*this.names.length)],
		lastName: this.names[Math.floor(Math.random()*this.names.length)]
	};

	return fullName;
};

NameGenerator.prototype.getRandomNameByGender = function(gender){
	if(gender == "male"){
		return this.maleNames[Math.floor(Math.random()*this.maleNames.length)];
	}
	else if(gender == "female"){
		return this.femaleName[Math.floor(Math.random()*this.femaleName.length)];
	}
};

NameGenerator.prototype.getRandomNamesByGender = function(number, gender){
	var randomNames = [];
	for (var i = 0; i < number; i++) {
		var randomName = null;
		if(gender == "male"){
			randomName = this.maleNames[Math.floor(Math.random()*this.maleNames.length)];
		}else if(gender == "female"){
			randomName = this.femaleName[Math.floor(Math.random()*this.femaleName.length)];
		}

		randomNames.push(randomName);
	}

	return randomNames;
};

NameGenerator.prototype.getFullNameByGender = function(gender){
	var randomNames = [];

	for (var i = 0; i < 3; i++) {
		var randomName = null;
		if(gender == "male"){
			randomName = this.maleNames[Math.floor(Math.random()*this.maleNames.length)];
		}
		else if(gender == "female"){
			randomName = this.femaleName[Math.floor(Math.random()*this.femaleName.length)];
		}
		
		randomNames.push(randomName);
	}

	return randomNames.join(" ");
};

NameGenerator.prototype.getFullNameObjectByGender = function(gender){
	var fullName = null;
	
	if(gender == "male"){
		fullName = {
			name: this.maleNames[Math.floor(Math.random()*this.maleNames.length)],
			firstName: this.maleNames[Math.floor(Math.random()*this.maleNames.length)],
			lastName: this.maleNames[Math.floor(Math.random()*this.maleNames.length)]
		};
	}
	else if(gender == "female"){
		fullName = {
			name: this.femaleName[Math.floor(Math.random()*this.femaleName.length)],
			firstName: this.femaleName[Math.floor(Math.random()*this.femaleName.length)],
			lastName: this.femaleName[Math.floor(Math.random()*this.femaleName.length)]
		};
	}
	

	return fullName;
};


module.exports = new NameGenerator();
